### DISCUSSION TITLE
Migrating from 3.4 to 5.x.x any horror stories ?

### DISCUSSION THREAD

**Discussion by oldnoob2** [phalcon/phalcon#295] [General]
Hello,

I'm migrating my code base again, this third time, first 1.x to 2.x then 2 to 3, I skip 4.x.  Now it is inevitable. Lots of composer packages requires minimum 7.4. I getting annoyed to find older versions of those packages.

So I want to upgrade my basement.

This time its from 3.4 to 5.x.x

I wonder is there any horror stories, warnings etc.

My Humble Regards.

**Comment by borisdelev** [comment]
Well... Depends of your project, but probably will be easy- changing namespaces, functions are renamed and a few deprecated stuff. And maybe the ->save() function... need to assign variables before that.

You can do it 😉

**Comment by niden** [comment]
Exactly what @borisdelev wrote. It depends on your project and how you have it structured.

Other than namespace changes and some functions renamed, the biggest hurdle you will have to overcome is the return type of the `findFirst()` if you use that. It is now `null` vs. `false`. As such any conditionals to check for `false` will fail.

Feel free to ask any questions you might have here or in Discord.

**Reply by oldnoob2** [reply]
Thanks, guys. Actually I shuddered a bit when I saw the forum empty. It seems everybody moved to discord.

It's a large monolith (more than 150k lines). So far main issue was paginator. I did very bad mistake in 3.4 (probably older than that) it used to work somehow, but now it needs to be changed. So I have to update lots of Model.

It looks like it will take a lot more time than I anticipated.

Phalcon continues to amaze me with every version update. And It's really amazing how backwards the Phalcon is.

**Comment by DavisDS** [comment]
Depends on how old the version you are on I suppose, but between v3 and 5 several namespaces changed, and with type safety in PHP 7 and 8 there is a lot of work to add variable types.  That said, the changes made in phalcon 5 make sense, and the new versions of PHP are critical if you are concerned with security.

**Comment by qmegas** [comment]
I have few projects written on version 3.4 and I'm thinking about the upgrade for long time, because I want to upgrade php to latest versions also. But I'm not sure if to do it right now and upgrade to version 5.x or to wait and upgrade to 6.x

**Reply by DavisDS** [reply]
6.x is going to be a pure PHP version.  This means less headache with installing extensions but at the cost of performance.  I think you'll be waiting a while on 6 to be fully ready and to match the performance of a current version

**Reply by qmegas** [reply]
Yes, I know. I like the idea that version 6.x will be pure php version and at the same time we will be able to install extension to improve performance. So I probably will wait patiently, I don't want to upgrade twice (now to version 5.x and then to 6.x when it will be ready).

**Comment by oldnoob2** [comment]
Hi,

I would like to state that the upgrade process has been completed successfully.

On phalcon side there was some namespace changes

It works more than 2 weeks and there is no problem. 

Even my code was not so standard.

Thanks for all support.

**Comment by fvromera** [comment]
Hi Guys, 

Only 2 important additions if you use Phalcon 5+ with php 8.2:

- callable: There is some problems with magic methods findByXXX/findFirstByXXX that can be solved avoiding this magic methods, or replacing with their equivalent using find/findFirst(['conditions' => ...])
- classes dynamic properties is deprecated: You can avoid it extending ``stdClass`` or adding ``#[\AllowDynamicProperties]`` above class definition

**Comment by shovv** [comment]
For everybody going through same process I suggest to check your project dependecies. If any of them is built using Phalcon 3 / 4 is better to be sure they are updated for Phalcon 5 support. Otherwise you might drown with workarounds.
